addons/mrp/models/mrp_production.py

```diff
@@ -1454,77 +1452,228 @@ def _get_name_backorder(self, name, sequence):
+                available_qty = move_line.reserved_uom_qty
+                available_qty = move_line.product_uom_id._compute_quantity(
+                    available_qty, product_uom)
+                if float_compare(available_qty, 0, precision_rounding=move_line.product_uom_id.rounding) <= 0:
+                    continue
+                ml_by_move.append((move_line.product_uom_id._compute_quantity(
+                    available_qty, product_uom), move_line, move_line.copy_data()[0]))
+
+            initial_move.move_line_ids.with_context(
+                bypass_reservation_update=True).write({'reserved_uom_qty': 0})
+            moves = list(initial_move | backorder_moves)
+
+            move = moves and moves.pop(0)
+            move_qty_to_reserve = move.product_qty
+            for quantity, move_line, ml_vals in ml_by_move:
+                while quantity and move:
+                    taken_qty = min(move_qty_to_reserve, quantity)
+                    taken_qty_uom = product_uom._compute_quantity(
+                        taken_qty, move_line.product_uom_id)
+                    if move == initial_move:
+                        move_line.with_context(
+                            bypass_reservation_update=True).reserved_uom_qty = taken_qty_uom
+                    else:
+                        move_lines_vals.append(dict(
+                            ml_vals,
+                            reserved_uom_qty=taken_qty_uom,
+                            move_id=move.id
+                        ))
+                    quantity -= taken_qty
+                    move_qty_to_reserve -= taken_qty
+
+                    if move_qty_to_reserve <= 0:
+                        move_to_assign.add(move.id)
+                        move = moves and moves.pop(0)
+                        move_qty_to_reserve = move and move.product_qty or 0
+
+                # Unreserve the quantity removed from initial `stock.move.line` and
+                # not assign to a move anymore. In case of a split smaller than initial
+                # quantity and fully reserve
+                if quantity:
+                    self.env['stock.quant']._update_reserved_quantity(
+                        move_line.product_id, move_line.location_id, -quantity,
+                        lot_id=move_line.lot_id, package_id=move_line.package_id,
+                        owner_id=move_line.owner_id, strict=True)
+
+            if move and move_qty_to_reserve != move.product_qty:
+                move_to_partialy_assign.add(move.id)
+
+            move_lines_to_unlink.update(initial_move.move_line_ids.filtered(
+                lambda ml: not ml.reserved_uom_qty and not ml.qty_done).ids)
+
+        self.env['stock.move'].browse(
+            move_to_assign).write({'state': 'assigned'})
+        self.env['stock.move'].browse(
+            move_to_partialy_assign).write({'state': 'partially_available'})
+        self.env['stock.move.line'].browse(move_lines_to_unlink).unlink()
```

Comment:
I would erase move_id on the move line you delete to avoid calling `recompute_state`